File: mainapp/Classes/Managers/VoiceMoldManager.cpp
//
// Created by Taihyun Hwang on 16-Mar-2018.
//

#if (CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID)

#include "VoiceMoldManager.h"

#include <jni.h>
#include "platform/android/jni/JniHelper.h"
#include "cocos2d.h"
#include <string>
#include <limits>

using namespace cocos2d;
using namespace std;

std::string VoiceMoldManager::getDefaultLocale(std::string locale)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticGetDefaultLocale",
                                       "()Ljava/lang/String;"))
    {
        jstring jLocale = (jstring)t.env->CallStaticObjectMethod(t.classID, t.methodID);
        string locale = JniHelper::jstring2string(jLocale);

        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jLocale);
        
        return locale;
    }
    
    return "";
}

void VoiceMoldManager::setDefaultLocale(std::string locale)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticSetDefaultLocale",
                                       "(Ljava/lang/String;)V"))
    {
        jstring jLocale = t.env->NewStringUTF(locale.c_str());
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jLocale);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jLocale);
    }
}

void VoiceMoldManager::secureVoiceData()
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticSecureVoiceData", "()V"))
    {
        t.env->CallStaticVoidMethod(t.classID, t.methodID);
        t.env->DeleteLocalRef(t.classID);
    }
}

void VoiceMoldManager::secureVoiceData(std::string locale)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticSecureVoiceData", "(Ljava/lang/String;)V"))
    {
        jstring jLocale = t.env->NewStringUTF(locale.c_str());
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jLocale);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jLocale);
    }
}

void VoiceMoldManager::warmup()
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticWarmup", "()V"))
    {
        t.env->CallStaticVoidMethod(t.classID, t.methodID);
        t.env->DeleteLocalRef(t.classID);
    }
}

void VoiceMoldManager::warmup(std::string locale)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticWarmup", "(Ljava/lang/String;)V"))
    {
        jstring jLocale = t.env->NewStringUTF(locale.c_str());
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jLocale);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jLocale);
    }
}

void VoiceMoldManager::speak(std::string text)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticSpeak", "(Ljava/lang/String;)V"))
    {
        jstring jText = t.env->NewStringUTF(text.c_str());
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jText);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jText);
    }
}

void VoiceMoldManager::speak(std::string text, std::string locale)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticSpeak", "(Ljava/lang/String;Ljava/lang/String;)V"))
    {
        jstring jText = t.env->NewStringUTF(text.c_str());
        jstring jLocale = t.env->NewStringUTF(locale.c_str());
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jText, jLocale);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jText);
        t.env->DeleteLocalRef(jLocale);
    }
}

float VoiceMoldManager::guessSpeakDuration(std::string text)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticGuessSpeakDuration", "(Ljava/lang/String;)F"))
    {
        jstring jText = t.env->NewStringUTF(text.c_str());
        float rv = t.env->CallStaticFloatMethod(t.classID, t.methodID, jText);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jText);
        
        return rv;
    }
    
    return std::numeric_limits<float>::quiet_NaN();
}

float VoiceMoldManager::guessSpeakDuration(std::string text, std::string locale)
{
    JniMethodInfo t;
    if (JniHelper::getStaticMethodInfo(t,
                                       "org/cocos2dx/cpp/VoiceMoldManager",
                                       "staticGuessSpeakDuration", "(Ljava/lang/String;Ljava/lang/String;)F"))
    {
        jstring jText = t.env->NewStringUTF(text.c_str());
        jstring jLocale = t.env->NewStringUTF(locale.c_str());
        float rv = t.env->CallStaticFloatMethod(t.classID, t.methodID, jText, jLocale);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jText);
        t.env->DeleteLocalRef(jLocale);
        
        return rv;
    }
    
    return std::numeric_limits<float>::quiet_NaN();
}

VoiceMoldManager* VoiceMoldManager::shared()
{
    static VoiceMoldManager* _shared = nullptr;
    if (!_shared) {
        _shared = new VoiceMoldManager();
    }
    return _shared;

}

#endif
//EOF

File: mainapp/proj.android-studio/app/src/org/cocos2dx/cpp/VoiceMoldManager.java
/**
 * VoiceMoldManager.java -- Thin wrapper for locale-to-tts mapping
 *
 * NB(xenosoz, 2018): Date created: 26 Feb, 2018
 */


package org.cocos2dx.cpp;
import android.content.Context;
import android.content.Intent;

import java.lang.String;
import java.util.HashMap;


public class VoiceMoldManager
{
    public String defaultLocale = "en-IN";
    public HashMap<String, VoiceMold> moldForLocaleCache;

    public String getDefaultLocale() { return defaultLocale; }
    public void setDefaultLocale(String locale) { defaultLocale = locale; }

    public VoiceMoldManager() {
        moldForLocaleCache = new HashMap<String, VoiceMold>();
    }

    public VoiceMold moldForLocale(String locale) {
        VoiceMold cache = moldForLocaleCache.get(locale);
        if (cache != null) {
            return cache;
        }

        VoiceMold mold = new VoiceMold(locale);
        moldForLocaleCache.put(locale, mold);
        return mold;
    }

    public void secureVoiceData() {
        secureVoiceData(defaultLocale);
    }

    public void secureVoiceData(String locale) {
        VoiceMold mold = moldForLocale(locale);

        if (mold != null && mold.isEnergetic()) {
            return;
        }

        Intent intent = new Intent();
        intent.setAction("com.enuma.voice_engine_a.installer.CALL_BY_KITKIT_SCHOOL");

        Context context = AppActivity.instance();
        context.startActivity(intent);
    }

    public void warmup() {
        warmup(defaultLocale);
    }

    public void warmup(String locale) {
        VoiceMold mold = moldForLocale(locale);
        mold.warmup();
    }

    public void speak(String text) {
        speak(text, defaultLocale);
    }

    public void speak(String text, String locale) {
        VoiceMold mold = moldForLocale(locale);
        mold.speak(text);
    }

    public float guessSpeakDuration(String text) {
        return guessSpeakDuration(text, defaultLocale);
    }

    public float guessSpeakDuration(String text, String locale) {
        VoiceMold mold = moldForLocale(locale);
        return mold.guessSpeakDuration(text);
    }


    // ----
    private static VoiceMoldManager _shared;

    public static VoiceMoldManager shared() {
        if (_shared == null) {
            _shared = new VoiceMoldManager();
        }
        return _shared;
    }


    // NB(xenosoz, 2018): Sugers for simplifying exposure to c++ world.

    public static String staticGetDefaultLocale() {
        return shared().getDefaultLocale();
    }
    public static void staticSetDefaultLocale(String locale) {
        shared().setDefaultLocale(locale);
    }

    public static void staticSecureVoiceData() { shared().secureVoiceData(); }
    public static void staticSecureVoiceData(String locale) { shared().secureVoiceData(locale); }

    public static void staticWarmup() { shared().warmup(); }
    public static void staticWarmup(String locale) { shared().warmup(locale); }

    public static void staticSpeak(String text) { shared().speak(text); }
    public static void staticSpeak(String text, String locale) {
        shared().speak(text, locale);
    }

    public static float staticGuessSpeakDuration(String text) { return shared().guessSpeakDuration(text); }
    public static float staticGuessSpeakDuration(String text, String locale) { return shared().guessSpeakDuration(text); }
}
